File: Java/src/main/java/connectn/ConnectNGameBuilder.java
package connectn;

/**
 * Builder for creating ConnectNGame instances with a fluent, readable API.
 * 
 * <p>This builder solves several problems with the multi-parameter constructor:
 * <ul>
 *   <li>Parameter order confusion (rows vs columns)</li>
 *   <li>Lack of default values</li>
 *   <li>Difficulty adding optional parameters</li>
 *   <li>Unclear constructor calls</li>
 *   <li>No way to create common configurations easily</li>
 * </ul>
 * 
 * <p>Example usage:
 * <pre>{@code
 * // Create a custom game with fluent API
 * ConnectNGame game = new ConnectNGameBuilder()
 *     .rows(6)
 *     .columns(7)
 *     .winLength(4)
 *     .build();
 * 
 * // Use preset configurations
 * ConnectNGame ticTacToe = ConnectNGameBuilder.ticTacToe();
 * ConnectNGame connectFour = ConnectNGameBuilder.connectFour();
 * 
 * // Create with defaults (3x3, win length 3)
 * ConnectNGame defaultGame = new ConnectNGameBuilder().build();
 * }</pre>
 * 
 * @author CS5004 Teaching Team
 * @version 1.0
 */
public class ConnectNGameBuilder {
    // Default values for common configurations
    private static final int DEFAULT_ROWS = 3;
    private static final int DEFAULT_COLUMNS = 3;
    private static final int DEFAULT_WIN_LENGTH = 3;
    
    // Builder fields with defaults
    private int rows = DEFAULT_ROWS;
    private int columns = DEFAULT_COLUMNS;
    private int winLength = DEFAULT_WIN_LENGTH;
    
    /**
     * Creates a new builder with default values (3x3 board, win length 3).
     */
    public ConnectNGameBuilder() {
        // Use default values
    }
    
    /**
     * Sets the number of rows for the game board.
     *
     * @param rows the number of rows (must be ≥ 1)
     * @return this builder for method chaining
     */
    public ConnectNGameBuilder rows(int rows) {
        this.rows = rows;
        return this;
    }
    
    /**
     * Sets the number of columns for the game board.
     *
     * @param columns the number of columns (must be ≥ 1)
     * @return this builder for method chaining
     */
    public ConnectNGameBuilder columns(int columns) {
        this.columns = columns;
        return this;
    }
    
    /**
     * Sets the win length (number of consecutive tokens needed to win).
     *
     * @param winLength the win length (must be ≥ 1 and ≤ min(rows, columns))
     * @return this builder for method chaining
     */
    public ConnectNGameBuilder winLength(int winLength) {
        this.winLength = winLength;
        return this;
    }
    
    /**
     * Sets the board dimensions (rows and columns) in one call.
     * 
     * <p>This is a convenience method for setting both dimensions at once.
     *
     * @param rows the number of rows (must be ≥ 1)
     * @param columns the number of columns (must be ≥ 1)
     * @return this builder for method chaining
     */
    public ConnectNGameBuilder boardSize(int rows, int columns) {
        this.rows = rows;
        this.columns = columns;
        return this;
    }
    
    /**
     * Creates a square board with the specified size.
     * 
     * <p>This is a convenience method for creating square boards (rows = columns).
     *
     * @param size the size of the square board (must be ≥ 1)
     * @return this builder for method chaining
     */
    public ConnectNGameBuilder squareBoard(int size) {
        this.rows = size;
        this.columns = size;
        return this;
    }
    
    /**
     * Builds and returns a new ConnectNGame instance with the configured parameters.
     * 
     * <p>This method delegates to the ConnectNGame constructor, which performs
     * all validation. If the parameters are invalid, the constructor will throw
     * an IllegalArgumentException.
     *
     * @return a new ConnectNGame instance
     * @throws IllegalArgumentException if any parameter is invalid
     */
    public ConnectNGame build() {
        return new ConnectNGame(rows, columns, winLength);
    }
    
    // ==================== Preset Factory Methods ====================
    
    /**
     * Creates a standard Tic-Tac-Toe game (3x3 board, win length 3).
     * 
     * <p>This is equivalent to:
     * <pre>{@code
     * new ConnectNGameBuilder()
     *     .rows(3)
     *     .columns(3)
     *     .winLength(3)
     *     .build();
     * }</pre>
     *
     * @return a new ConnectNGame configured for Tic-Tac-Toe
     */
    public static ConnectNGame ticTacToe() {
        return new ConnectNGameBuilder()
            .rows(3)
            .columns(3)
            .winLength(3)
            .build();
    }
    
    /**
     * Creates a standard Connect Four game (6x7 board, win length 4).
     * 
     * <p>This is equivalent to:
     * <pre>{@code
     * new ConnectNGameBuilder()
     *     .rows(6)
     *     .columns(7)
     *     .winLength(4)
     *     .build();
     * }</pre>
     *
     * @return a new ConnectNGame configured for Connect Four
     */
    public static ConnectNGame connectFour() {
        return new ConnectNGameBuilder()
            .rows(6)
            .columns(7)
            .winLength(4)
            .build();
    }
    
    /**
     * Creates a Gomoku game (15x15 board, win length 5).
     * 
     * <p>Gomoku is a traditional Japanese board game also known as Five in a Row.
     * This is equivalent to:
     * <pre>{@code
     * new ConnectNGameBuilder()
     *     .rows(15)
     *     .columns(15)
     *     .winLength(5)
     *     .build();
     * }</pre>
     *
     * @return a new ConnectNGame configured for Gomoku
     */
    public static ConnectNGame gomoku() {
        return new ConnectNGameBuilder()
            .rows(15)
            .columns(15)
            .winLength(5)
            .build();
    }
    
    /**
     * Creates a small practice game (5x5 board, win length 4).
     * 
     * <p>This configuration is good for quick games and testing.
     *
     * @return a new ConnectNGame configured for a small practice game
     */
    public static ConnectNGame smallGame() {
        return new ConnectNGameBuilder()
            .rows(5)
            .columns(5)
            .winLength(4)
            .build();
    }
    
    /**
     * Creates a large strategic game (10x10 board, win length 5).
     * 
     * <p>This configuration provides more strategic depth and longer gameplay.
     *
     * @return a new ConnectNGame configured for a large strategic game
     */
    public static ConnectNGame largeGame() {
        return new ConnectNGameBuilder()
            .rows(10)
            .columns(10)
            .winLength(5)
            .build();
    }
    
    /**
     * Creates a custom square board game with the specified size and win length.
     * 
     * <p>This is a convenience factory method for creating square boards.
     *
     * @param size the size of the square board (must be ≥ 1)
     * @param winLength the win length (must be ≥ 1 and ≤ size)
     * @return a new ConnectNGame with a square board
     * @throws IllegalArgumentException if parameters are invalid
     */
    public static ConnectNGame customSquare(int size, int winLength) {
        return new ConnectNGameBuilder()
            .squareBoard(size)
            .winLength(winLength)
            .build();
    }
}

File: Python/connectn/connect_n_game_builder.py
"""Builder for creating ConnectNGame instances with a fluent, readable API."""

from typing import Optional
from .connect_n_game import ConnectNGame


class ConnectNGameBuilder:
    """
    Builder for creating ConnectNGame instances with a fluent, readable API.
    
    This builder solves several problems with the multi-parameter constructor:
    - Parameter order confusion (rows vs columns)
    - Lack of default values
    - Difficulty adding optional parameters
    - Unclear constructor calls
    - No way to create common configurations easily
    
    Example usage:
        >>> # Create a custom game with fluent API
        >>> game = (ConnectNGameBuilder()
        ...         .rows(6)
        ...         .columns(7)
        ...         .win_length(4)
        ...         .build())
        >>> 
        >>> # Use preset configurations
        >>> tic_tac_toe = ConnectNGameBuilder.tic_tac_toe()
        >>> connect_four = ConnectNGameBuilder.connect_four()
        >>> 
        >>> # Create with defaults (3x3, win length 3)
        >>> default_game = ConnectNGameBuilder().build()
    """
    
    # Default values for common configurations
    DEFAULT_ROWS = 3
    DEFAULT_COLUMNS = 3
    DEFAULT_WIN_LENGTH = 3
    
    def __init__(self):
        """Create a new builder with default values (3x3 board, win length 3)."""
        self._rows = self.DEFAULT_ROWS
        self._columns = self.DEFAULT_COLUMNS
        self._win_length = self.DEFAULT_WIN_LENGTH
    
    def rows(self, rows: int) -> 'ConnectNGameBuilder':
        """
        Set the number of rows for the game board.
        
        Args:
            rows: The number of rows (must be >= 1)
        
        Returns:
            This builder for method chaining
        """
        self._rows = rows
        return self
    
    def columns(self, columns: int) -> 'ConnectNGameBuilder':
        """
        Set the number of columns for the game board.
        
        Args:
            columns: The number of columns (must be >= 1)
        
        Returns:
            This builder for method chaining
        """
        self._columns = columns
        return self
    
    def win_length(self, win_length: int) -> 'ConnectNGameBuilder':
        """
        Set the win length (number of consecutive tokens needed to win).
        
        Args:
            win_length: The win length (must be >= 1 and <= min(rows, columns))
        
        Returns:
            This builder for method chaining
        """
        self._win_length = win_length
        return self
    
    def board_size(self, rows: int, columns: int) -> 'ConnectNGameBuilder':
        """
        Set the board dimensions (rows and columns) in one call.
        
        This is a convenience method for setting both dimensions at once.
        
        Args:
            rows: The number of rows (must be >= 1)
            columns: The number of columns (must be >= 1)
        
        Returns:
            This builder for method chaining
        """
        self._rows = rows
        self._columns = columns
        return self
    
    def square_board(self, size: int) -> 'ConnectNGameBuilder':
        """
        Create a square board with the specified size.
        
        This is a convenience method for creating square boards (rows = columns).
        
        Args:
            size: The size of the square board (must be >= 1)
        
        Returns:
            This builder for method chaining
        """
        self._rows = size
        self._columns = size
        return self
    
    def build(self) -> ConnectNGame:
        """
        Build and return a new ConnectNGame instance with the configured parameters.
        
        This method delegates to the ConnectNGame constructor, which performs
        all validation. If the parameters are invalid, the constructor will raise
        a ValueError.
        
        Returns:
            A new ConnectNGame instance
        
        Raises:
            ValueError: If any parameter is invalid
        """
        return ConnectNGame(self._rows, self._columns, self._win_length)
    
    # ==================== Preset Factory Methods ====================
    
    @staticmethod
    def tic_tac_toe() -> ConnectNGame:
        """
        Create a standard Tic-Tac-Toe game (3x3 board, win length 3).
        
        This is equivalent to:
            ConnectNGameBuilder().rows(3).columns(3).win_length(3).build()
        
        Returns:
            A new ConnectNGame configured for Tic-Tac-Toe
        """
        return ConnectNGameBuilder().rows(3).columns(3).win_length(3).build()
    
    @staticmethod
    def connect_four() -> ConnectNGame:
        """
        Create a standard Connect Four game (6x7 board, win length 4).
        
        This is equivalent to:
            ConnectNGameBuilder().rows(6).columns(7).win_length(4).build()
        
        Returns:
            A new ConnectNGame configured for Connect Four
        """
        return ConnectNGameBuilder().rows(6).columns(7).win_length(4).build()
    
    @staticmethod
    def gomoku() -> ConnectNGame:
        """
        Create a Gomoku game (15x15 board, win length 5).
        
        Gomoku is a traditional Japanese board game also known as Five in a Row.
        This is equivalent to:
            ConnectNGameBuilder().rows(15).columns(15).win_length(5).build()
        
        Returns:
            A new ConnectNGame configured for Gomoku
        """
        return ConnectNGameBuilder().rows(15).columns(15).win_length(5).build()
    
    @staticmethod
    def small_game() -> ConnectNGame:
        """
        Create a small practice game (5x5 board, win length 4).
        
        This configuration is good for quick games and testing.
        
        Returns:
            A new ConnectNGame configured for a small practice game
        """
        return ConnectNGameBuilder().rows(5).columns(5).win_length(4).build()
    
    @staticmethod
    def large_game() -> ConnectNGame:
        """
        Create a large strategic game (10x10 board, win length 5).
        
        This configuration provides more strategic depth and longer gameplay.
        
        Returns:
            A new ConnectNGame configured for a large strategic game
        """
        return ConnectNGameBuilder().rows(10).columns(10).win_length(5).build()
    
    @staticmethod
    def custom_square(size: int, win_length: int) -> ConnectNGame:
        """
        Create a custom square board game with the specified size and win length.
        
        This is a convenience factory method for creating square boards.
        
        Args:
            size: The size of the square board (must be >= 1)
            win_length: The win length (must be >= 1 and <= size)
        
        Returns:
            A new ConnectNGame with a square board
        
        Raises:
            ValueError: If parameters are invalid
        """
        return ConnectNGameBuilder().square_board(size).win_length(win_length).build()
